warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/handler/edge/ProtobufEdgeDataTypeHandler.java

```diff
@@ -634,45 +638,73 @@ public long process(KEYIN key, RawRecordContainer event, Multimap<String,Normali
+                        }
+                    }
+                }
+                // Otherwise we want to multiplex across all values for the sink and source regardless of subgroup
+                else {
+                    for (String sourceSubGroup : mSource.keySet()) {
+                        for (NormalizedContentInterface ifaceSource : mSource.get(sourceSubGroup)) {
+                            for (String sinkSubGroup : mSink.keySet()) {
+                                for (NormalizedContentInterface ifaceSink : mSink.get(sinkSubGroup)) {
+                                    EdgeDataBundle edgeValue = createEdge(edgeDef, event, ifaceSource, sourceGroup, sourceSubGroup, ifaceSink, sinkGroup,
+                                                    sinkSubGroup, edgeAttribute2, edgeAttribute3, normalizedFields, depthFirstList, loadDateStr, activityDate,
+                                                    validActivityDate);
+                                    if (edgeValue != null) {
+                                        
+                                        // have to write out the keys as the edge values are generated, so counters get updated
+                                        // and the system doesn't timeout.
+                                        edgesCreated += writeEdges(edgeValue, context, contextWriter, validActivityDate, activityEqualsEvent, event.getDate());
+                                        
+                                        if (this.enableMetadata) {
+                                            registerEventMetadata(eventMetadataRegistry, enrichmentFieldName, edgeValue, jexlPreconditions);
+                                        }
+                                    }
                                 }
                             }
-                            
                         }
                     }
                 }
-            }
-            // Otherwise we want to multiplex across all values for the sink and source regardless of subgroup
-            else {
-                for (String sourceSubGroup : mSource.keySet()) {
-                    for (NormalizedContentInterface ifaceSource : mSource.get(sourceSubGroup)) {
-                        for (String sinkSubGroup : mSink.keySet()) {
-                            for (NormalizedContentInterface ifaceSink : mSink.get(sinkSubGroup)) {
-                                EdgeDataBundle edgeValue = createEdge(edgeDef, event, ifaceSource, sourceGroup, sourceSubGroup, ifaceSink, sinkGroup,
-                                                sinkSubGroup, edgeAttribute2, edgeAttribute3, normalizedFields, depthFirstList, loadDateStr, activityDate,
-                                                validActivityDate);
+                
+            } else {
+                if (sourceGroup.equals(sinkGroup) && (sourceGroup != NO_GROUP)) {
+                    Set<String> commonKeys = matchingGroups.get(sourceGroup);
```

Comment:
I think this is the same loop as above.  A separate function perhaps?  That leads us to perhaps creating separate functions from each section.  This method is getting a little too large.  That would also help with unit testing each piece.

Reply:
Yep, will do.

Reply:
After beginning this refactor, it's become obvious that this class needs some massive TLC.  Created issue 1776 for this and leaving as is for now.